man/corosync-keygen.8
.\"/*
.\" * Copyright (C) 2010-2019 Red Hat, Inc.
.\" *
.\" * All rights reserved.
.\" *
.\" * Author: Angus Salkeld <asalkeld@redhat.com>
.\" *
.\" * This software licensed under BSD license, the text of which follows:
.\" *
.\" * Redistribution and use in source and binary forms, with or without
.\" * modification, are permitted provided that the following conditions are met:
.\" *
.\" * - Redistributions of source code must retain the above copyright notice,
.\" *   this list of conditions and the following disclaimer.
.\" * - Redistributions in binary form must reproduce the above copyright notice,
.\" *   this list of conditions and the following disclaimer in the documentation
.\" *   and/or other materials provided with the distribution.
.\" * - Neither the name of the MontaVista Software, Inc. nor the names of its
.\" *   contributors may be used to endorse or promote products derived from this
.\" *   software without specific prior written permission.
.\" *
.\" * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
.\" * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
.\" * THE POSSIBILITY OF SUCH DAMAGE.
.\" */
.TH COROSYNC-KEYGEN 8 2019-04-09
.SH NAME
corosync-keygen \- Generate an authentication key for Corosync.
.SH SYNOPSIS
.B "corosync-keygen [\-k <filename>]  [-m <randomfile>] [\-s size] [\-l] [\-h]"
.SH DESCRIPTION

If you want to configure corosync to use cryptographic techniques to ensure authenticity
and privacy of the messages, you will need to generate a private key.
.PP
.B corosync-keygen
creates this key and writes it to /etc/corosync/authkey or to file specified by
-k option.
.PP
This private key must be copied to every processor in the cluster.  If the
private key isn't the same for every node, those nodes with nonmatching private
keys will not be able to join the same configuration.
.PP
Copy the key to some security transportable storage or use ssh to transmit the
key from node to node.  Then install the key with the command:
.PP
unix#: install -D --group=0 --owner=0 --mode=0400 /path_to_authkey/authkey /etc/corosync/authkey
.PP
If a message "Invalid digest" appears from the corosync executive, the keys
are not consistent between processors.
.PP
.SH OPTIONS
.TP
.B -k <filename>
This specifies the fully qualified path to the shared key to create.
.br
The default is /etc/corosync/authkey.
.TP
.B -r
Random number source file. Default is /dev/urandom. As an example /dev/random may be
used when really superb randomness is needed.
.TP
.B -s size
Size of the generated key in bytes. Default is 256 bytes. Allowed range is <128, 4096>.
.TP
.TP
.B -l
Option is not used and it's kept only for compatibility.
.TP
.B -h
Print basic usage.

.SH EXAMPLES
.TP
Generate the key.
.nf
# corosync-keygen
Corosync Cluster Engine Authentication key generator.
Gathering 2048 bits for key from /dev/urandom.
Writing corosync key to /etc/corosync/authkey
.fi

.TP
Generate longer key and store it in the /tmp/authkey file.
.nf
$ corosync-keygen -s 2048 -k /tmp/authkey
Corosync Cluster Engine Authentication key generator.
Gathering 16384 bits for key from /dev/urandom.
Writing corosync key to /tmp/authkey.
.fi

.TP
Generate superb key using /dev/random
.nf
# corosync-keygen -r /dev/random
Gathering 2048 bits for key from /dev/random.
Press keys on your keyboard to generate entropy.
Press keys on your keyboard to generate entropy (1128 bits still needed).
Press keys on your keyboard to generate entropy (504 bits still needed).
Press keys on your keyboard to generate entropy (128 bits still needed).
Press keys on your keyboard to generate entropy (32 bits still needed).
Writing corosync key to /etc/corosync/authkey.
.fi

.SH SEE ALSO
.BR corosync_overview (7),
.BR corosync.conf (5),
.SH AUTHOR
Angus Salkeld
.PP


CHANGELOG
------------------------------------------------------------------------
r3035 | sdake | 2010-08-31 12:04:23 -0700 (Tue, 31 Aug 2010) | 6 lines

Merge trunk revision 3034:
r3034 | sdake | 2010-08-31 12:02:19 -0700 (Tue, 31 Aug 2010) | 2 lines

Bump version to 1.2.8.


------------------------------------------------------------------------
r3033 | sdake | 2010-08-31 12:01:51 -0700 (Tue, 31 Aug 2010) | 7 lines

Merge trunk revision 3030:
r3030 | sdake | 2010-08-31 11:55:27 -0700 (Tue, 31 Aug 2010) | 3 lines

Build properly with different versions of libc by including all posix APIs
in header definitions.


------------------------------------------------------------------------
r3032 | sdake | 2010-08-31 12:01:27 -0700 (Tue, 31 Aug 2010) | 6 lines

Merge trunk revision 3029:
r3029 | asalkeld | 2010-08-24 18:13:14 -0700 (Tue, 24 Aug 2010) | 2 lines

IPC: return CS_ERR_NO_RESOURCES to library when low on fds.


------------------------------------------------------------------------
r3031 | sdake | 2010-08-31 12:00:52 -0700 (Tue, 31 Aug 2010) | 6 lines

Merge trunk revision 3028:
r3028 | asalkeld | 2010-08-24 18:07:37 -0700 (Tue, 24 Aug 2010) | 2 lines
    
POLL: gracefully handle running out of file descriptors.


------------------------------------------------------------------------
r3027 | sdake | 2010-08-24 13:28:46 -0700 (Tue, 24 Aug 2010) | 5 lines

Merge trunk revision 3024:
r3024 | sdake | 2010-08-24 11:45:43 -0700 (Tue, 24 Aug 2010) | 2 lines

Remove checking of sub parameters in service.d files.

------------------------------------------------------------------------
r3026 | sdake | 2010-08-24 13:09:54 -0700 (Tue, 24 Aug 2010) | 9 lines

Merge trunk revision 3011:
r3011 | honzaf | 2010-08-02 05:38:19 -0700 (Mon, 02 Aug 2010) | 5 lines
    
Check that Corosync was not runned by cman
    
Patch fixes init script so it's tested, if corosync was runned
by cman or not. If so, it refuses to stop Corosync.


------------------------------------------------------------------------
r3025 | sdake | 2010-08-24 13:07:05 -0700 (Tue, 24 Aug 2010) | 11 lines

Merge trunk revision 3010:
r3010 | honzaf | 2010-08-02 05:36:20 -0700 (Mon, 02 Aug 2010) | 7 lines

Allow running only one instance of Corosync

Patch makes Corosync more compliant with common practices
for writing daemon. It creates pid file (usually
/var/run/corosync.pid) and flocks it. So only one instance
of Corosync can be executed now.


------------------------------------------------------------------------
r3023 | sdake | 2010-08-17 11:11:25 -0700 (Tue, 17 Aug 2010) | 6 lines

Merge trunk revision 3022:
r3022 | sdake | 2010-08-17 11:08:13 -0700 (Tue, 17 Aug 2010) | 2 lines

Properly detect shutdown of corosync process


------------------------------------------------------------------------
r3013 | sdake | 2010-08-03 10:36:37 -0700 (Tue, 03 Aug 2010) | 6 lines

Merge trunk revision 3012:
r3012 | sdake | 2010-08-03 10:31:33 -0700 (Tue, 03 Aug 2010) | 2 lines

Remove cancel token retransmit timeout.


------------------------------------------------------------------------
r3008 | sdake | 2010-07-27 14:37:46 -0700 (Tue, 27 Jul 2010) | 6 lines

Merge trunk revision 3007:
r3007 | sdake | 2010-07-27 14:32:39 -0700 (Tue, 27 Jul 2010) | 2 lines

Change trunk version to 1.2.7.


------------------------------------------------------------------------
r3006 | sdake | 2010-07-27 12:08:29 -0700 (Tue, 27 Jul 2010) | 7 lines

Merge trunk revision 3005:
r3005 | sdake | 2010-07-27 12:00:37 -0700 (Tue, 27 Jul 2010) | 3 lines

Remove consensus check for two node cluster cases which can have smaller
consensus values.  Document in man page the behavior of consensus.


------------------------------------------------------------------------
r3004 | asalkeld | 2010-07-21 16:16:41 -0700 (Wed, 21 Jul 2010) | 5 lines

Merge trunk revision 3002:
r3002 | sdake | 2010-07-22 06:48:40 +1000 (Thu, 22 Jul 2010) | 2 lines
Fix merge error with revision 3001.


------------------------------------------------------------------------
r3003 | asalkeld | 2010-07-21 16:13:42 -0700 (Wed, 21 Jul 2010) | 6 lines

Merge trunk revision 3001:
r3001 | sdake | 2010-07-22 03:03:36 +1000 (Thu, 22 Jul 2010) | 3 lines
Fix problem where flow control could lock up ipc under very heavy load in very
rare circumstances.


------------------------------------------------------------------------
r3000 | sdake | 2010-07-19 13:20:25 -0700 (Mon, 19 Jul 2010) | 12 lines

Merge trunk revision 2995:
r2995 | asalkeld | 2010-07-16 21:59:40 -0700 (Fri, 16 Jul 2010) | 9 lines

SYNC: always call sync_aborted() in sync_confchg_fn().

1) sync_callbacks.sync_abort can be null.
2) sync_processing is set to 0 after syncv1 is done.
   Then syncv2 processing is down. If we get a config change
   after syncv1 is down, but before syncv2 is done then it won't
   get aborted.


------------------------------------------------------------------------
r2999 | sdake | 2010-07-19 13:19:30 -0700 (Mon, 19 Jul 2010) | 6 lines

Merge trunk revision 2991:
r2991 | asalkeld | 2010-07-15 19:08:54 -0700 (Thu, 15 Jul 2010) | 2 lines

 SYNCV2: reset the my_memb_determine_ring_id in sync_v2_memb_list_abort()


------------------------------------------------------------------------
r2998 | sdake | 2010-07-19 08:29:30 -0700 (Mon, 19 Jul 2010) | 7 lines

Merge trunk revision 2997:
r2997 | fabbione | 2010-07-18 23:36:48 -0700 (Sun, 18 Jul 2010) | 4 lines

Fix logging_daemon config parser code.

Resolves: rhbz#615203

------------------------------------------------------------------------
r2989 | sdake | 2010-07-14 11:36:53 -0700 (Wed, 14 Jul 2010) | 8 lines

Merge trunk revision 2988:
r2988 | sdake | 2010-07-14 11:35:36 -0700 (Wed, 14 Jul 2010) | 4 lines

Remove reset of token timeout on retransmitted token reception.  The timer
should only be reset when a real token is received or membership protocol
could run into problems with certain timing parameters.



tools/corosync-keygen.c
/*
 * Copyright (c) 2004 MontaVista Software, Inc.
 * Copyright (c) 2005-2019 Red Hat, Inc.
 *
 * All rights reserved.
 *
 * Author: Steven Dake (sdake@redhat.com)
 *         Jan Friesse (jfriesse@redhat.com)
 *
 * This software licensed under BSD license, the text of which follows:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * - Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 * - Neither the name of the MontaVista Software, Inc. nor the names of its
 *   contributors may be used to endorse or promote products derived from this
 *   software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config.h>

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/stat.h>

#include <netinet/in.h>

#include <corosync/totem/totem.h>

#define DEFAULT_KEYFILE COROSYSCONFDIR "/authkey"

#define DEFAULT_KEYFILE_LEN		256

#define DEFAULT_RANDOM_DEV		"/dev/urandom"

static const char usage[] =
	"Usage: corosync-keygen [-k <keyfile>] [-s size] [-m <randomfile>] [-l] [-h]\n"
	"     -k / --key-file=<filename> -  Write to the specified keyfile\n"
	"            instead of the default " DEFAULT_KEYFILE ".\n"
	"     -r / --random-file -  Random number source file. Default is \n"
	"            /dev/urandom. As an example /dev/random may be requested\n"
	"            (that may require user input for entropy).\n"
	"     -l / --less-secure - Not used, option is kept only\n"
	"            for compatibility.\n"
	"     -s / --size -  Length of key.\n"
	"     -h / --help -  Print basic usage.\n";


int main (int argc, char *argv[])
{
	int authkey_fd;
	int random_fd;
	char *keyfile = NULL;
	unsigned char key[TOTEM_PRIVATE_KEY_LEN_MAX];
	ssize_t res;
	ssize_t bytes_read;
	size_t key_len = DEFAULT_KEYFILE_LEN;
	const char *random_dev = DEFAULT_RANDOM_DEV;
	long long int tmpll;
	char *ep;
	int c;
	int option_index;
	static struct option long_options[] = {
		{ "key-file",    required_argument, NULL, 'k' },
		{ "less-secure", no_argument,       NULL, 'l' },
		{ "random-file", required_argument, NULL, 'r' },
		{ "size",        required_argument, NULL, 's' },
		{ "help",        no_argument,       NULL, 'h' },
		{ 0,             0,                 NULL, 0   },
	};

	while ((c = getopt_long (argc, argv, "k:r:s:lh",
			long_options, &option_index)) != -1) {
		switch (c) {
		case 'k':
			keyfile = optarg;
			break;
		case 'l':
			/*
			 * Only kept for compatibility
			 */
			break;
		case 'r':
			random_dev = optarg;
			break;
		case 's':
			tmpll = strtoll(optarg, &ep, 10);
			if (tmpll < TOTEM_PRIVATE_KEY_LEN_MIN ||
			    tmpll > TOTEM_PRIVATE_KEY_LEN_MAX ||
			    errno != 0 || *ep != '\0') {
				errx (1, "Unsupported key size (supported <%u,%u>)\n",
				    TOTEM_PRIVATE_KEY_LEN_MIN,
				    TOTEM_PRIVATE_KEY_LEN_MAX);
			}

			key_len = (size_t)tmpll;
			break;
		case 'h':
			printf ("%s\n", usage);
			exit(0);
			break;
		default:
			printf ("Error parsing command line options.\n");
			exit (1);
		}
	}

	printf ("Corosync Cluster Engine Authentication key generator.\n");

	if (!keyfile) {
		keyfile = (char *)DEFAULT_KEYFILE;
	}

	printf ("Gathering %lu bits for key from %s.\n", (unsigned long)(key_len * 8), random_dev);
	random_fd = open (random_dev, O_RDONLY);

	if (random_fd == -1) {
		err (1, "Failed to open random source");
	}

	if (strcmp(random_dev, "/dev/random") == 0) {
		printf ("Press keys on your keyboard to generate entropy.\n");
	}
	/*
	 * Read random data
	 */
	bytes_read = 0;

retry_read:
	res = read (random_fd, &key[bytes_read], key_len - bytes_read);
	if (res == -1) {
		err (1, "Could not read /dev/random");
	}
	bytes_read += res;
	if (bytes_read != key_len) {
		printf ("Press keys on your keyboard to generate entropy (%d bits still needed).\n",
		    (int)((key_len - bytes_read) * 8));
		goto retry_read;
	}
	close (random_fd);

	/*
	 * Open key
	 */
	authkey_fd = open (keyfile, O_CREAT|O_WRONLY|O_TRUNC, 0600);
	if (authkey_fd == -1) {
		err (2, "Could not create %s", keyfile);
	}
	if (fchmod (authkey_fd, 0400)) {
		err (3, "Failed to set key file permissions to 0400");
	}

	printf ("Writing corosync key to %s.\n", keyfile);

	/*
	 * Write key
	 */
	res = write (authkey_fd, key, key_len);
	if (res != key_len) {
		err (4, "Could not write %s", keyfile);
	}

	if (close (authkey_fd)) {
		err (5, "Could not close %s", keyfile);
	}

	return (0);
}
